File: src/java/ExpireBin.java
//Copyright 2014 Aerospike, Inc
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import com.aerospike.client.AerospikeClient;
import com.aerospike.client.AerospikeException;
import com.aerospike.client.Key;
import com.aerospike.client.Language;
import com.aerospike.client.Log;
import com.aerospike.client.Record;
import com.aerospike.client.ScanCallback;
import com.aerospike.client.Value;
import com.aerospike.client.policy.Policy;
import com.aerospike.client.policy.ScanPolicy;
import com.aerospike.client.policy.WritePolicy;
import com.aerospike.client.query.Statement;
import com.aerospike.client.task.ExecuteTask;
import com.aerospike.client.task.RegisterTask;


public class ExpireBin {
	static final String MODULE_NAME = "expire_bin";
	public AerospikeClient client;
	public WritePolicy policy;

	/**
	 * Initialize ExpireBin object with suitable client and policy
	 * 
	 * @param client - client to perform operations on
	 * @param policy - configuration parameters for expire bin ops
	 */
	public ExpireBin(AerospikeClient client, WritePolicy policy) {
		this.policy = policy;
		this.client = client;
	}

	/**
	 * Try to get values from the expire bin
	 * 
	 * @param - key to get from
	 * @param bins -  list of bin names to attempt to get from
	 * @return list containing respective values from each bin, or null if
	 * expired or bin doesn't exist
	 * @throws AerospikeException 
	 */
	public Object get(Key key, String ... bins) throws AerospikeException {
		Value[] valueBins = new Value[bins.length];
		int count = 0;
		for (String bin : bins) {
			valueBins[count] = Value.get(bin);
			count++;
		}
		return client.execute(policy, key, MODULE_NAME, "get", valueBins);
	}

	/**
	 * Create or update expire bins. If the expCreate flag is set to true, 
	 * all newly created bins will be expire bins otherwise, only normal bins will be created
	 * and existing expire bins will be updated. Note: existing expire bins will not be converted
	 * into normal bins if create flag is off.
	 * 
	 * @param key - Record key to apply operation on
	 * @param binName - Bin name to create or update
	 * @param val - bin value
	 * @param binTTL - expiration time in seconds or -1 for no expiration
	 * @param expCreate - set true to create expire bins, false to create normal bins
	 * @return 0 in success, 1 if error
	 * @throws AerospikeException 
	 */
	public Integer put(Key key, String binName, Value val, int binTTL, boolean expCreate) throws AerospikeException {
		int create;
		if (expCreate == false) {
			create = 1;
		}else {
			create = 0;
		}
		return (Integer) client.execute(policy, key, MODULE_NAME, "put", Value.get(binName), val, Value.get(binTTL), Value.get(create));
	}

	/**
	 * Batch create or update expire bins for a given key. Use the PutBin map
	 * wrapper to store each put operation. Don't supply a binTTL to turn bin 
	 * creation off. 
	 * 
	 * @param key - Record key to store bins
	 * @param mapBins - list of PutBins containing operation arguments
	 * @return - 0 if all bins succeeded, 1 if failure
	 * @throws AerospikeException 
	 */
	public Object puts(Key key, PutBin ... mapBins) throws AerospikeException {
		Value[] valueMaps = new Value[mapBins.length];
		int count = 0;
		for (PutBin map : mapBins) {
			valueMaps[count] = Value.get(map);
			count++;
		}
		return client.execute(policy, key, MODULE_NAME, "puts", valueMaps);
	}

	/**
	 * Batch update the bin TTLs. Use this method to change or reset 
	 * the bin TTL of multiple bins in a record. 
	 * 
	 * @param key - Record key
	 * @param mapBins - list of TouchBins containing operation arguments
	 * @return 0 on success of all touch operations, 1 if a failure occurs
	 * @throws AerospikeException 
	 */
	public Object touch(Key key, TouchBin ... mapBins) throws AerospikeException {
		Value[] valueMaps = new Value[mapBins.length];
		int count = 0;
		for (Map map : mapBins) {
			valueMaps[count] = Value.get(map);
			count++;
		}
		return client.execute(policy, key, MODULE_NAME, "touch", valueMaps);
	}

	/**
	 * Perform a scan of the database and clear out expired expire bins
	 * 
	 * @param scan - scan policy containing which records should be scanned
	 * @param namespace - namespace of server to scan
	 * @param set - set of server to scan
	 * @param bins - list of bins to scan
	 * @throws AerospikeException 
	 */
	public ExecuteTask clean(Policy policy, Statement statement, String ... bins) throws AerospikeException {
		final Value[] valueBins = new Value[bins.length];
		int count = 0;
		for (String bin : bins) {
			valueBins[count] = Value.get(bin);
			count++;
		}
		return client.execute(policy, statement, MODULE_NAME, "clean", valueBins);
	}

	/**
	 * Get time bin will expire in seconds.
	 * 
	 * @param key - Record key
	 * @param bin - Bin Name
	 * @return time in seconds bin will expire, -1 or null if it doesn't expire
	 * @throws AerospikeException 
	 */
	public Integer ttl(Key key, String bin) throws AerospikeException {
		return (Integer)(client.execute(policy, key, MODULE_NAME, "ttl", Value.get(bin)));
	}

	public class PutBin extends HashMap<Value, Value> {
		/**
		 * Map wrapper to store batch put operations
		 */
		private static final long serialVersionUID = 1L;

		
		/**
		 * Map to create / update a bin with bin creation turned on
		 * 
		 * @param binName
		 * @param val
		 * @param binTTL
		 */
		public PutBin(String binName, Value val, int binTTL) {
			super();
			this.put(Value.get("bin"), Value.get(binName));
			this.put(Value.get("bin_ttl"), Value.get(binTTL));
			this.put(Value.get("val"), val);
		}
		
		/**
		 * Map to create / update a bin with bin creation turned off
		 * 
		 * @param binName
		 * @param val
		 */
		public PutBin(String binName, Value val) {
			super();
			this.put(Value.get("bin"), Value.get(binName));
			this.put(Value.get("val"), val);
		}
	}
	

	public class TouchBin extends HashMap<Value, Value> {
		/**
		 * Map wrapper to store touch operations
		 */
		private static final long serialVersionUID = 1L;
		
		/**
		 * Map to perform a touch bin operation
		 * 
		 * @param binName - name of bin to touch
		 * @param binTTL - new TTL for bin
		 */
		public TouchBin(String binName, int binTTL) {
			super();
			this.put(Value.get("bin"), Value.get(binName));
			this.put(Value.get("bin_ttl"), Value.get(binTTL));
		}
	}
	
	public static void main(String[] args) {
		AerospikeClient testClient = null;
		System.out.println("This is a demo of the expirable bin module for Java");
		try {
			System.out.println("Connecting to server...");
			testClient = new AerospikeClient("127.0.0.1", 3000);
			System.out.println("Connected!");

			ExpireBin eb = new ExpireBin(testClient, new WritePolicy());
			System.out.println("Creating expire bins...");
			Key testKey1 = null, testKey2 = null, testKey3 = null;
			testKey1 = new Key("test", "expireBin", "eb1");
			testKey2 = new Key("test", "expireBin", "eb2");
			testKey3 = new Key("test", "expireBin", "eb3");

			System.out.println(((Integer)(eb.put(testKey1, "TestBin", Value.get("Hello World"), -1, true)) == 0 ? "TestBin 1 inserted." : "TestBin 1 not inserted" ));
			System.out.println((Integer)eb.put(testKey2, "TestBin", Value.get("I don't expire"), -1, true) == 0 ? "TestBin 2 inserted" : "TestBin 2 not inserted");
			System.out.println((Integer)eb.put(testKey3, "TestBin", Value.get("I will expire soon"), 5, true) == 0 ? "TestBin 3 inserted" : "TestBin 3 not inserted");

			System.out.println("Getting expire bins...");

			System.out.println("TestBin 1: " + eb.get(testKey1, "TestBin").toString());
			System.out.println("TestBin 2: " + eb.get(testKey2, "TestBin").toString());
			System.out.println("TestBin 3: " + eb.get(testKey3, "TestBin").toString());

			System.out.println("Current time is: " + (System.currentTimeMillis() / 1000) + "\nGetting bin TTLs...");
			System.out.println("TestBin 1 TTL: " + eb.ttl(testKey1, "TestBin"));
			System.out.println("TestBin 2 TTL: " + eb.ttl(testKey2, "TestBin"));
			System.out.println("TestBin 3 TTL: " + eb.ttl(testKey3, "TestBin"));


			System.out.println("Waiting for TestBin 3 to expire...");

			try {
				TimeUnit.SECONDS.sleep(10);
			} catch(InterruptedException ex) {
				Thread.currentThread().interrupt();
			}

			System.out.println("Getting expire bins again...");

			System.out.println("TestBin 1: " + eb.get(testKey1, "TestBin").toString());
			System.out.println("TestBin 2: " + eb.get(testKey2, "TestBin").toString());
			System.out.println("TestBin 3: " + eb.get(testKey3, "TestBin").toString());

			System.out.println("Cleaning bins...");
			Statement stmt = new Statement();
			stmt.setNamespace("test");
			stmt.setSetName("expireBin");
			eb.clean(new WritePolicy(), stmt, "TestBin");
		} catch (AerospikeException e) {
			e.printStackTrace();
			System.exit(1);
		}



	}

}

File: src/python/expire_bin.py
#!/usr/bin/env python

# Copyright 2014 Aerospike, Inc

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#    http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import aerospike
import time

MODULE_NAME = "expire_bin"
GET_OP = "get"
PUT_OP = "put"
BATCH_PUT_OP = "puts"
TTL_OP = "ttl"
TOUCH_OP = "touch"
CLEAN_OP = "clean"
CITRUSLEAF_EPOCH = 1262304000

class ExpireBin:
	def __init__(self, client):
		"""Initialize the ExpireBin module

		Args:
			client -- aerospike client to perform ops
		"""

		self.client = client

	def get(self, policy, key, *bins):
		"""Attempt to retrieve values from list of bins. The bins
		can be expire bins or normal bins.

		Args:
			policy -- policy to use for op
			key -- tuple (namespace, set, record name)
			*bins -- one or more bin names to retrieve values from

		Returns:
			record dict: A record with each bin value mapped to the bin name.
			If the bin value expired or does not exist, there will not
			be an entry in the dict

		Raises:
			Exception: Exception with details of server error.
		"""
		rv = self.client.apply(key, MODULE_NAME, GET_OP, list(bins), policy)
		if not type(rv) == dict:
			raise Exception("Get operation failed, return {0} instead of map".format(type(rv)))
		return rv

	def put(self, policy, key, bin, val, bin_ttl):
		"""Create or update expire bins. If bin_ttl is not None,
		all newly created bins will be expire bins otherwise, only normal bins will be created
		Note: existing expire bins will not be converted into normal bins if bin_ttl is None.

		Args:
			policy -- policy to use for op
			key -- tuple (namespace, set, record name)
			bin -- bin name
			val -- value to store into bin
			bin_ttl -- expiration time in seconds or -1 for no expiration

		Returns:
			int: 0 if success, 1 otherwise

		Raises:
			Exception: Exception with details of server error.
		"""
		return self.client.apply(key, MODULE_NAME, PUT_OP, [bin, val, bin_ttl], policy)

	def puts(self, policy, key, *binMaps):
		"""Batch create or update expire bins for a given key. Use the dict
			{'bin' : bin_name, 'val' : bin_value, 'bin_ttl' : ttl}
		to store each put operation. Omit the bin_ttl to turn bin creation off.

		Args:
			policy -- policy to use for op
			key -- tuple (namespace, set, record name)
			*binMaps -- dict {'bin' : bin_name, 'val' : bin_val, 'bin_ttl' : ttl}

		Returns:
			int: 0 if all ops succeeded, 1 if error occurred

		Raises:
			Exception: Exception with details of server error.

		"""
		return self.client.apply(key, MODULE_NAME, BATCH_PUT_OP, list(binMaps), policy)

	def touch(self, policy, key, *mapBins):
		"""Batch update the bin TTLs. Us this method to change or reset the bin TTL of
		multiple bins in a record. Use a dict to store each touch operation.

		Args:
			policy -- policy to use for op
			key -- tuple (namespace, set, record name)
			*mapBins -- dict {'bin' : bin_name, 'bin_ttl' : ttl}

		Returns:
			int: 0 on success of ALL touch operations, 1 if a failure occurs

		Raises:
			Exception: Exception with details of server error.
		"""
		return self.client.apply(key, MODULE_NAME, TOUCH_OP, list(mapBins), policy)

	def clean(self, policy, scan, *bins):
		"""Clear out the expired bins on a scan of the database

		Args:
			policy -- policy to use for op
			scan -- Scan object to run bin clean on
			*bins -- bin names to clean out

		Raises:
			Exception: Exception with details of server error.
		"""
		raise NotImplementedError("Scan UDF not yet implemented in python client")
		#def callback((key, meta, record)):
		#	self.client.apply(key, MODULE_NAME, "does_not_exist", list(bins), self.policy)
		#scan.foreach(callback)

	def ttl(self, policy, key, bin):
		"""Get the time bin will expire in seconds.

		Args:
			policy -- policy to use for op
			key -- tuple (namespace, set, record name)
			bin -- bin name

		Returns:
			int: Time to expire in seconds

		Raises:
			Exception: Exception with details of server error.
		"""
		return self.client.apply(key, MODULE_NAME, TTL_OP, [bin], policy)

def main():
	config = { 'hosts' : [ ('127.0.0.1', 3000) ]}
	policy = { "timeout" : 2000 }
	testClient = aerospike.client(config).connect()
	try:
		print "Registering UDF..."
		filename = "../../expire_bin.lua"
		udf_type = 0
		testClient.udf_put(policy, filename, udf_type)
		print "UDF Registered!"
	except Exception as e:
		print "Error registering lua function: {0}".format(e)

	eb = ExpireBin(testClient)
	key = ("test", "expireBin", "eb")

	print "Creating expire bins..."

	print "TestBin 1: {0}".format("Inserted!" if eb.put(policy, key, "TestBin1", "Hello World", -1) is 0 else "Insertion Failed")
	print "TestBin 2: {0}".format("Inserted!" if eb.put(policy, key, "TestBin2", "I don't expire", -1) is 0 else "Insertion Failed")
	print "TestBin 3: {0}".format("Inserted!" if eb.put(policy, key, "TestBin3", "I will expire soon", 5 ) is 0 else "Insertion Failed")
	print "TestBin 4 & 5: {0}".format("Inserted!" if eb.puts(policy, key, {'bin' : "TestBin4", 'val' : "Good Morning.", 'bin_ttl' : 100}, {'bin' : "TestBin5", 'val' : "Good Night."})  is 0 else "Insertion Failed")


	print "Getting expire bins..."

	print "TestBins: {0}".format(eb.get(policy, key, "TestBin1", "TestBin2", "TestBin3", "TestBin4", "TestBin5"))

	print "Current time is: {0}\nGetting bin TTLs...".format(time.time() - CITRUSLEAF_EPOCH)
	print "TestBin 1 TTL: {0}".format(eb.ttl(policy, key, "TestBin1"))
	print "TestBin 2 TTL: {0}".format(eb.ttl(policy, key, "TestBin2"))
	print "TestBin 3 TTL: {0}".format(eb.ttl(policy, key, "TestBin3"))
	print "TestBin 4 TTL: {0}".format(eb.ttl(policy, key, "TestBin4"))
	print "TestBin 5 TTL: {0}".format(eb.ttl(policy, key, "TestBin5"))


	print "Waiting for TestBin 3 to expire..."

	time.sleep(10)

	print "Getting expire bins again..."

	print "TestBins: {0}".format(eb.get(policy, key, "TestBin1", "TestBin2", "TestBin3", "TestBin4", "TestBin5"))

	print "Changing expiration times..."

	eb.touch(policy, key, {'bin' : "TestBin1", 'bin_ttl' : 10}, {'bin' : "TestBin4", 'bin_ttl' : 5});

	print "Getting bin TTLs again..."

	print "TestBin 1 TTL: {0}".format(eb.ttl(policy, key, "TestBin1"))
	print "TestBin 2 TTL: {0}".format(eb.ttl(policy, key, "TestBin2"))
	print "TestBin 3 TTL: {0}".format(eb.ttl(policy, key, "TestBin3"))
	print "TestBin 4 TTL: {0}".format(eb.ttl(policy, key, "TestBin4"))
	print "TestBin 5 TTL: {0}".format(eb.ttl(policy, key, "TestBin5"))


	# scan udf not yet implemented in python client
	print "Cleaning bins..."

	testScan = testClient.scan("test", "expireBin")
	try:
		eb.clean(policy, testScan, "TestBin1", "TestBin2", "TestBin3", "TestBin4", "TestBin5")
	except NotImplementedError as e:
		print "Scan UDF not implemented in current version of python client."

	testClient.close()
if __name__ == "__main__":
	main()


